File: 2022/Day3.hs
#! /usr/bin/env -S stack runhaskell
import Data.List
import Data.List.Split
import Data.Char

type Input = [String]
type Output = Int

parse :: String -> Input
parse = lines

parts :: Int -> [a] -> [[a]]
parts k xs = chunksOf (length xs `div` k) xs

prio :: Char -> Int
prio c
  | isLower c = ord c - ord 'a' + 1
  | otherwise = ord c - ord 'A' + 27

sumCommonPriorities :: [[String]] -> Int
sumCommonPriorities = sum . map (prio . head . foldr1 intersect)

part1 :: Input -> Output
part1 = sumCommonPriorities . map (parts 2)

part2 :: Input -> Output
part2 = sumCommonPriorities . chunksOf 3

main :: IO()
main = do
    f <- readFile("3.in")
    let input = parse f
    putStr "Solution for part 1: "
    putStrLn . show $ part1 input
    putStr "Solution for part 2: "
    putStrLn . show $ part2 input

-- Solution part 1: 8139
-- Solution part 2: 2668

File: 2022/Day3.py
#! /usr/bin/env python
# pylint: disable=unused-wildcard-import
from util import *

def parse(inp):
    return lines(inp)

def prio(c):
    if ord(c) >= ord('a') and ord(c) <= ord('z'):
        return ord(c) - ord('a') + 1
    if ord(c) >= ord('A') and ord(c) <= ord('Z'):
        return ord(c) - ord('A') + 27
    return 0

def intersectRucksack(rucksack):
    return set.intersection(*(map(set, parts(rucksack, 2)))).pop()

def intersectBadge(badge):
    return set.intersection(*map(set, badge)).pop()

def p1(inp):
    return sum(prio(intersectRucksack(rucksack)) for rucksack in inp)

def p2(inp):
    return sum(prio(intersectBadge(badge)) for badge in chunks(inp, 3))

def main():
    inp = parse(data())
    print(f"Solution for part 1:\n{p1(inp)}")
    print(f"Solution for part 2:\n{p2(inp)}")

if __name__ == "__main__":
    main()

# Solution part 1: 8139
# Solution part 2: 2668
# Leaderboard: 882 / 605
